pallets/market/src/benchmarking.rs

```diff
@@ -0,0 +1,409 @@
+    let price_per_block: u32 = 10;
+    let cost = (end_block - start_block) * price_per_block * n;
+
+    let mut proposals = vec![];
+    for _ in 1..=n {
+        let label = vec![0; MAX_LABEL_SIZE as usize];
+
+        let proposal = DealProposal::<T::AccountId, BalanceOf<T>, BlockNumberFor<T>> {
+            piece_cid: random_cid(&mut small_rng).to_bytes().try_into().unwrap(),
+            piece_size: 2048,
+            client: client.clone(),
+            provider: caller.clone(),
+            label: BoundedVec::try_from(label).unwrap(),
+            start_block: start_block.into(),
+            end_block: end_block.into(),
+            storage_price_per_block: price_per_block.into(),
+            provider_collateral: COLLATERAL.into(),
+            state: DealState::<_>::Published,
+        };
+
+        let signed_deal_proposal =
+            create_ed25519_signature(proposal.encode().as_slice(), client_pair.clone()).into();
+
+        let proposal = ClientDealProposal::<T::AccountId, BalanceOf<T>, BlockNumberFor<T>, _> {
+            proposal,
+            client_signature: signed_deal_proposal,
+        };
+
+        proposals.push(proposal);
+    }
+    return (proposals, cost);
+}
+
+#[frame_benchmarking::v2::benchmarks(
+    where
+        T: crate::Config<OffchainSignature = sp_runtime::MultiSignature>,
+        T: pallet_balances::Config,
+        T: pallet_storage_provider::Config<
+            PeerId = BoundedPeerIdBytes,
+        >,
+        T: frame_system::Config<AccountId = AccountId32>
+)]
+mod benchmarks {
+    use itertools::Itertools;
+
+    use super::*;
+
+    const EXISTENTIAL_DEPOSIT: u32 = 1_000_000_000;
+
+    fn setup_account_balance<T>(account: T::AccountId)
+    where
+        T: crate::Config,
+        T: pallet_balances::Config,
+    {
+        pallet_balances::Pallet::<T>::make_free_balance_be(
+            &account,
+            (EXISTENTIAL_DEPOSIT * 2).into(),
+        );
+
+        // Add some balance so we can withdraw it
+        MarketPallet::<T>::add_balance(
+            RawOrigin::Signed(account.clone()).into(),
+            EXISTENTIAL_DEPOSIT.into(),
+        )
+        .unwrap();
+
+        assert_eq!(
+            MarketPallet::<T>::free(&account).unwrap(),
+            EXISTENTIAL_DEPOSIT.into()
+        );
+        assert_eq!(MarketPallet::<T>::locked(&account).unwrap(), 0u32.into());
+    }
+
+    /// Run until a particular block.
+    ///
+    /// Stolen't from: <https://github.com/paritytech/polkadot-sdk/blob/7df94a469e02e1d553bd4050b0e91870d6a4c31b/substrate/frame/lottery/src/mock.rs#L87-L98>
+    pub fn run_to_block<T>(n: u32)
+    where
+        T: crate::Config,
+        T: pallet_storage_provider::Config,
+    {
+        use frame_support::traits::Hooks;
+
+        while frame_system::Pallet::<T>::block_number() < n.into() {
+            if frame_system::Pallet::<T>::block_number() > 1u32.into() {
+                pallet_storage_provider::Pallet::<T>::on_finalize(
```

Comment:
Starting to wonder whether we should mock out the `pallet_storage_provider`. I mean, we're only using it here because `StorageProviderValidation` trait.

Reply:
raise this again if a similar pattern comes up again, for now - not worth it